File: c/adc/read.c
// Read the voltage values from the 
// built-in ADC
// http://www.acmesystems.it/?id=foxg20_adc

#include "stdio.h"
#include "stdlib.h"
#include "fcntl.h"
#include "unistd.h"
#include "sys/types.h"
#include "sys/stat.h"
#include "sys/ioctl.h"
#include "math.h"

#define ADC_SYSFS "/sys/bus/platform/devices/at91_adc/chan"
#define VREF 3.24
#define V_PER_POINT VREF/2**10

int ADC_Read(int channel){
	int fd,rtc;
	char filename[41];
	char buffer[5];

	sprintf(filename,"%s%d",ADC_SYSFS,channel);

	if ((fd=open(filename,O_RDONLY))<0) return -1;

	read(fd,buffer,5);
	rtc = atoi(buffer);
	close(fd);

	return rtc;
}


void main(void) {
	int channel;

	while(1) {
		for (channel=0;channel<4;channel++) {
			printf("Channel %d = %.2f\n",channel,(float)ADC_Read(channel)*VREF/pow(2,10));
		}
		printf("-------\n");
		sleep(1);
	} 

}

File: python/adc/read.py
#!/usr/bin/python

#Read the voltage values from the 
#built-in ADC
#http://www.acmesystems.it/foxg20_adc
 
import time 

VREF=3.24
volt_per_point=VREF/2**10
path="/sys/bus/platform/devices/at91_adc/"

ch=0
while True:

	fd = open(path + "chan" + str(ch),"r")
	sample = fd.read()
	print "Channel %d = %.2f volt" % (ch,int(sample)*volt_per_point)
	fd.close()
	time.sleep(0.5)
	ch+=1
	if ch==4: 
		ch=0
		print "-------"


